flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/PreciseTimestamp.java

```diff
@@ -0,0 +1,234 @@
+		this.milliseconds = milliseconds;
+		this.nanoseconds = nanoseconds;
+		this.precision = precision;
+	}
+
+	@Override
+	public int compareTo(PreciseTimestamp that) {
+		return this.milliseconds == that.milliseconds ?
+			Integer.compare(this.nanoseconds, that.nanoseconds) :
+			Long.compare(this.milliseconds, that.milliseconds);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (!(obj instanceof PreciseTimestamp)) {
+			return false;
+		}
+		PreciseTimestamp that = (PreciseTimestamp) obj;
+		return this.compareTo(that) == 0;
+	}
+
+	@Override
+	public String toString() {
+		// yyyy-MM-dd HH:mm:ss
+		StringBuilder timestampString = new StringBuilder(unixTimestampToString(milliseconds));
+		if (precision > 0 && nanoseconds > 0) {
+			// append fraction part
+			timestampString.append('.').append(nonosSecondsToString(nanoseconds, precision));
+		}
+		return timestampString.toString();
+	}
+
+	public long toLong() {
+		return milliseconds;
+	}
+
+	public static PreciseTimestamp fromLong(long milliseconds, int precision) {
+		Timestamp t = new Timestamp(milliseconds);
+		long millis = isCompact(precision) ?
+			zeroLastNDigits(t.getTime(), 3 - precision) : t.getTime();
+		int nanos = (int) zeroLastNDigits((long) t.getNanos(), 9 - precision);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public Timestamp toTimestamp() {
+		return Timestamp.valueOf(this.toString());
+	}
+
+	public static PreciseTimestamp fromTimestamp(Timestamp t, int precision) {
+		String timestampString = t.toString();
+		long millis = timestampStringToUnixDate(timestampString);
+		int nanos = timestampStringToNanoseconds(timestampString);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public LocalDateTime toLocalDateTime() {
+		long epochSeconds = (milliseconds >= 0 || nanoseconds == 0) ?
+			(milliseconds / 1000) : (milliseconds / 1000 - 1);
+		return LocalDateTime.ofEpochSecond(epochSeconds, nanoseconds, ZoneOffset.UTC);
+	}
+
+	public static PreciseTimestamp fromLocalDateTime(LocalDateTime ldt, int precision) {
+		int nanos = (int) zeroLastNDigits((long) ldt.getNano(), 9 - precision);
+		long millis = unixTimestamp(
+			ldt.getYear(),
+			ldt.getMonthValue(),
+			ldt.getDayOfMonth(),
+			ldt.getHour(),
+			ldt.getMinute(),
+			ldt.getSecond()) + (long) (nanos / 1_000_000);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public byte[] toUnscaledBytes() {
+		byte[] bytes8 = new byte[8];
+		long l = milliseconds;
+		for (int i = 0; i < 8; i++) {
+			bytes8[7 - i] = (byte) l;
+			l >>>= 8;
+		}
+
+		if (isCompact()) {
+			return bytes8;
+		}
+
+		byte[] bytes12 = new byte[12];
+		System.arraycopy(bytes8, 0, bytes12, 0, 8);
+		int n = nanoseconds;
+		for (int i = 0; i < 4; i++) {
+			bytes12[11 - i] = (byte) n;
+			n >>>= 8;
+		}
+
+		return bytes12;
+	}
+
+	public static PreciseTimestamp fromUnscaledBytes(byte[] bytes, int precision) {
```

Comment:
There is no need to convert between bytes and Timestamp.
We can write long and int directly.

Reply:
BinaryRow need this to write a PreciseTimestamp with precision > 3.

Reply:
You can directly use getMills and getNanos to write to `BinaryRow`.
Use `byte[]`, will have bad performance.

Reply:
OK, would write a long and a int directly.